adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala

```diff
@@ -123,6 +126,139 @@ trait GenomicRDD[T, U <: GenomicRDD[T, U]] {
+    require(sorted, "Cannot evenly repartition an unsorted RDD.")
+    val count = rdd.count
+    // we don't want a bunch of empty partitions, so we will just use count in
+    // the case the user wants more partitions than rdd records.
+    val finalPartitionNumber = Math.min(count, partitions)
+    // the average number of records on each node will help us evenly repartition
+    val average: Double = count.toDouble / finalPartitionNumber
+
+    // we already have a sorted rdd, so let's just move the data
+    // but we want to move it in blocks to maintain order
+    // zipWithIndex seems to be the cheapest way to guarantee this
+    val finalPartitionedRDD = rdd
+      .mapPartitionsWithIndex((idx, iter) => {
+        // here we get the destination partition number
+        // this will calculate so that each partition has a nearly
+        // equal amount of data on it
+        extractReferenceRegionsForPartition(iter, partitionMap.get(idx))
+      }, preservesPartitioning = true).zipWithIndex
+      .mapPartitionsWithIndex((idx, iter) => {
+        // divide the global index by the average to get the destination
+        // partition number
+        iter.map(_.swap).map(f => ((f._2._1, (f._1 / average).toInt), f._2._2))
+      }, preservesPartitioning = true)
+      .repartitionAndSortWithinPartitions(new ReferenceRegionRangePartitioner(finalPartitionNumber.toInt))
+
+    finalPartitionedRDD.cache()
+
+    val tempPartitionMap = finalPartitionedRDD.mapPartitions(iter =>
+      getRegionBoundsFromPartition(iter.map(f => (f._1._1, f._2))), preservesPartitioning = true).collect
+
+    // we want to try to avoid any None values in our final partition map
+    val newPartitionMap = tempPartitionMap.zipWithIndex.map(f => {
+      if (f._1.isEmpty && f._2 != 0) {
+        // in order to prevent None's, we just use the previous element's
+        // upper bound to build this partition bound.
+        Some({
+          val previousBound = tempPartitionMap(f._2 - 1).get._2
+          (
+            ReferenceRegion(previousBound.referenceName,
+              previousBound.start + 1,
+              previousBound.end + 1),
+              ReferenceRegion(previousBound.referenceName,
+                previousBound.start + 2,
+                previousBound.end + 2))
+        })
+      } else {
+        f._1
+      }
+    })
+
+    replaceRdd(finalPartitionedRDD.values, Some(newPartitionMap))
+  }
+
+  /**
+   * Extracts the correct reference region for each T given sorted order.
+   *
+   * @note A given T may map to multiple ReferenceRegions.
+   * @param iter The iterator (that partition's data) of T objects
+   * @param partitionMap the bounds of the partition where _._1 is the lower bound and _._2 is the
+   *                     upper bound
+   * @return An iterator of all T's keyed with the specific ReferenceRegion that the T represents in
+   *         the global sort
+   */
+  private def extractReferenceRegionsForPartition(
```

Comment:
This is useful for the case that a given tuple maps to multiple `ReferenceRegions`, otherwise it is a more complexj/inefficient way of getting `ReferenceRegions` out.